pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java

```diff
@@ -139,8 +142,22 @@ public static Optional<RawMessage> rebatchMessage(RawMessage msg,
 
                 ByteBuf metadataAndPayload = Commands.serializeMetadataAndPayload(Commands.ChecksumType.Crc32c,
                                                                                   metadata, compressedPayload);
-                Optional<RawMessage> result = Optional.of(new RawMessageImpl(msg.getMessageIdData(),
-                                                                             metadataAndPayload));
+
+                if (brokerEntryMetadata != null) {
+                    int brokerMetaSize = brokerEntryMetadata.getSerializedSize();
+                    ByteBuf brokerMeta =
+                            PulsarByteBufAllocator.DEFAULT.buffer(brokerMetaSize + 6, brokerMetaSize + 6);
+                    brokerMeta.writeShort(Commands.magicBrokerEntryMetadata);
+                    brokerMeta.writeInt(brokerMetaSize);
+                    brokerEntryMetadata.writeTo(brokerMeta);
+
+                    CompositeByteBuf compositeByteBuf = PulsarByteBufAllocator.DEFAULT.compositeBuffer();
```

Comment:
At first glance, can we make it to be direct memory? Since this relates to the message number, It should be more pressure on GC.

Reply:
Oh, it just affects the compacted message. The message number shouldn't be significant. It's fine here.